### PR TITLE
Adding template checking in CardTemplateEditor.java to solve issue #10431

### LINKED ISSUE #10431: If I accidentally enter the wrong field in "Card Types", all cards will show up empty
###### Reproduction Steps

1. Add some cards
2. Change their card types
3. An incorrect field is entered
4. When I studied the cards, I found the cards were empty

###### Expected Result

I should be prompted when I modify the field I entered

###### Actual Result

At the least, the cards should not appear empty

###### Video
ezgif com-gif-maker

###### Debug info

AnkiDroid Version = 2.16alpha45
Android Version = 11
Manufacturer = OnePlus
Model = KB2000
Hardware = qcom
Webview User Agent = Mozilla/5.0 (Linux; Android 11; KB2000 Build/RP1A.201005.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/87.0.4280.141 Mobile Safari/537.36
ACRA UUID = fdd28bd9-b444-4d17-8736-d76574185b51
Scheduler = std2
Crash Reports Enabled = true
DatabaseV2 Enabled = true

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by L-keeey**
## Purpose / Description
Field names will be check when attempt to create new card. But the template editing of created card is lack of checking field names, which will lead to the empty card problem. 

## Fixes
Fixes #10431

## Approach
The solution is add a field names checking when saving a template.

## How Has This Been Tested?

Through manual multiple tests, this scheme is verified to be feasible. By running local bundled test cases, only failures due to "Can't find dependent libraries" exist.

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 1]
```diff
@@ -676,13 +682,41 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+                }
+
+                if (modelHasChanged() && isFMTValid) {
                     View confirmButton = mTemplateEditor.findViewById(R.id.action_confirm);
                     if (confirmButton != null) {
                         if (!confirmButton.isEnabled()) {
                             Timber.d("CardTemplateEditor::discarding extra click after button disabled");
                             return true;
                         }
+
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:719]: revert this whitespace

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 2]
```diff
@@ -676,13 +682,41 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+                        if (!fieldNames.contains(fmt)){
+                            isFMTValid = false;
+                            break;
+                        }
+                    }
+                    }catch (Exception e){
+                        isFMTValid = false;
+                        Timber.e(e);
+                    }
+                }
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:710]: I think it's better to convert this code into a function and call that in the below if statement

**Review by Akshay0701 — CHANGES_REQUESTED** [review]
@L-keeey Thanks for the contribution,

please give a proper title to your PR, and I am also interested to show a message/dialog box to user, if they entered the wrong field names, what do you think?

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 3]
```diff
@@ -676,13 +682,41 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+
+                        if (!fieldNames.contains(fmt)){
+                            isFMTValid = false;
+                            break;
+                        }
+                    }
+                    }catch (Exception e){
+                        isFMTValid = false;
+                        Timber.e(e);
+                    }
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:709]:
```suggestion
                    try {
                    	Matcher matcher = Pattern.compile("\\{\\{([^\\{\\}])\\}\\}").matcher(tempAFMT + tempQFMT);

                    	while (matcher.find()) {
                        	String fmt = matcher.group(1);
                        	Timber.i("CardTemplateEditor:: FMT %s", fmt);

                        	if (!fieldNames.contains(fmt)) {
                            		isFMTValid = false;
                            		break;
                        	}
                    	}
                    } catch (Exception e) {
                    	isFMTValid = false;
                    	Timber.e(e);
                    }
```

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java (hunk 2 above)
Comment (L-keeey) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:710]: Thanks for the reply! I would convert it into a function later. About feature of showing a message/dialog box to user, I think my understanding of the project right now may not support adding the feature (also i may not have enough time to do it now). I think maybe you could make this feature as a new issue?
By the way, i failed in the code coverage test for the pull request. I'm wondering where should i add test cases for the new code?
Thanks for the reply again.

**Comment by L-keeey** [comment]
@Akshay0701 I have extracted the code into a function. This issue originally did not force to add the dialog response feature, which I considered too complicated for me. Could you please commit this version first? The checks are all passed now.

**Comment by L-keeey** [comment]
As I try to add dialog in the extracted method, the apps will directly exit template editing mode and left the dialog behind. Do you have any suggest? @Akshay0701

**Comment by mikehardy** [comment]
Updated description to correctly link (in special GitHub style) to the issue

```
## Fixes
Fixes #10431
```

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 4]
```diff
@@ -167,7 +175,6 @@ protected void onCreate(Bundle savedInstanceState) {
         startLoadingCollection();
     }
 
-
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:170]: can you revert this change please

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 5]
```diff
@@ -604,7 +611,6 @@ public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
                 menu.findItem(R.id.action_add_deck_override).setVisible(false);
             } else {
                 JSONObject template = getCurrentTemplate();
-
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:607]: revert

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 6]
```diff
@@ -704,6 +710,37 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+            JSONObject template = getCurrentTemplate();
+
+            if (template != null && mTemplateEditor.getFieldNames() != null){
+                String tempAFMT = template.getString("afmt");
+                String tempQFMT = template.getString("qfmt");
+
+                List<String> fieldNames = new ArrayList<>(mTemplateEditor.getFieldNames());
+                fieldNames.add("FrontSide");
+                try{
+                    Matcher matcher = Pattern.compile("\\{\\{([^\\{\\}]*)\\}\\}").matcher(tempQFMT+tempAFMT);
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:724]:
```suggestion
                    Matcher matcher = Pattern.compile("\\{\\{([^\\{\\}]*)\\}\\}").matcher(tempQFMT + tempAFMT);
```
i would suggest you to go through [code style](https://github.com/ankidroid/Anki-Android/wiki/Code-style)

**Review by Akshay0701 — CHANGES_REQUESTED** [review]
minor changes

**Comment by Akshay0701** [comment]
> This issue originally did not force to add the dialog response feature, which I considered too complicated for me. Could you please commit this version first? The checks are all passed now.

It's okay if you find it hard to implement,  

I stated I am interested in showing a dialog to let users know that they made a mistake while editing, it's just a suggestion and it is open for discussion on whether to include it or not.

**Comment by mikehardy** [comment]
> It's okay if you find it hard to implement,

Yep, and in fact, it's important to recognize when that's the case. There's lots of stuff I'm just not good at or not experienced enough at, and knowing that and being okay with it is a great skill, so in a general sense: I respect this stance a lot :clap: 

> I stated I am interested in showing a dialog to let users know that they made a mistake while editing, it's just a suggestion and it is open for discussion on whether to include it or not.

If we want a dialog (or if we want to discuss it) then @krmanik can you peel that out to a separate feature request open for discussion and we'll keep this PR focused on just checking + failure modes + error handlnig (but no dialog)?

**Comment by mikehardy** [comment]
Oops! Wrong tag, sorry @krmanik - above comment was for @Akshay0701 :point_up: :facepalm:

**Comment by Akshay0701** [comment]
Sure Mike, I will create a feature issue to discuss it.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 7]
```diff
@@ -1012,6 +1055,9 @@ private void addNewTemplate(JSONObject model) {
             // Set up question & answer formats
             newTemplate.put("qfmt", oldTemplate.getString("qfmt"));
             newTemplate.put("afmt", oldTemplate.getString("afmt"));
+            Timber.i("Invoke addNewTemplate(), %s", newTemplate);
+            System.out.println("get qfmt" + oldTemplate.getString("qfmt"));
+            System.out.println("get afmt" + oldTemplate.getString("afmt"));
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:1060]: use `Timber`

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 8]
```diff
@@ -704,6 +712,41 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+            JSONObject template = getCurrentTemplate();
+
+            if (template != null && mTemplateEditor.getFieldNames() != null){
+                String tempAFMT = template.getString("afmt");
+                String tempQFMT = template.getString("qfmt");
+
+                List<String> fieldNames = new ArrayList<>(mTemplateEditor.getFieldNames());
+                fieldNames.add("FrontSide");
+                try{
+                    Matcher matcher = Pattern.compile("\\{\\{([^\\{\\}]*)\\}\\}").matcher(tempQFMT + tempAFMT);
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:730]: How does this handle `:` and other field formstters?

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 9]
```diff
@@ -704,6 +712,41 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+        private boolean isFormatValid() {
+            boolean valid = true;
+            JSONObject template = getCurrentTemplate();
+
+            if (template != null && mTemplateEditor.getFieldNames() != null){
+                String tempAFMT = template.getString("afmt");
+                String tempQFMT = template.getString("qfmt");
+
+                List<String> fieldNames = new ArrayList<>(mTemplateEditor.getFieldNames());
+                fieldNames.add("FrontSide");
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:728]: * This is not the full list of field names: https://github.com/ankidroid/Anki-Android/blob/86db612ccc7ae4d2aff794cc64d7a06cd6fcdd0d/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java#L1140-L1171
* This is defined by code which is not controlled by us
* We should not fully block the user from saving the template: users may have a reason to save an invalid template (the above, or desktop addons)

Ideally we should be rendering the card and seeing if there is a problem, rather than duplicating the rendering logic as this will be 100% in-sync with the field requirements

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java (hunk 9 above)
Comment (L-keeey) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:728]: I'm sorry but I am not quite clear about your feedback. As @BrayanDSO said in original issue, all proper field names should be stored in the "mFieldNames" variable. The only special case is the name "FrontSide" which is not stored in "mFieldNames" but only used as a temporary symbol for a string replacement operation in function flipQA(). 
Also, if user really need to have other field names, I think they could've done this with a insert field action.

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:728]: My comment there was incomplete @L-keeey, I truly apologize for it.

I don't know how used you are to the Anki ecosystem, but there is some other keywords that Anki/AnkiDroid parses as well, such as `FrontSide`, `Tags` and the others you can see on the code mentioned by David.

There is also some "prefixes" (I don't know if it's appropriate it this way) that can be added to the fields, like `cloze:` and `edit:` (for [Edit Field During Review Addon](https://ankiweb.net/shared/info/1020366288)).

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java (hunk 8 above)
Comment (L-keeey) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:730]: Sorry but i am wondering where could i find all field formatters and their formatting rules? Since neither in the issue's description nor the target java file did i find the formatters. 
The situation now for me is i only know that with double `{}` the wrong field name will trigger the bug. If it is convenient for you, would you mind telling me what the rules are and would those rules also trigger the bug?
